### ISSUE TITLE
DataParallel does not correctly handle running variables in batch norm

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#1051]
Running variables accumulated on non-master GPUs are lost, and in the next step running variables accumulated on the master GPU are broadcast to non-master GPUs. Convergence-wise it does not make much difference;-), but it is still a bug. 
Also, device_ids in DataParallel should be checked to not contain the same values, otherwise stuff like this will break: 
https://github.com/pytorch/pytorch/blob/master/torch/backends/cudnn/rnn.py#L38-L43
If creating replicas on the same GPU is actually intended to be supported, then the above referenced lines that create a dictionary indexed by current_device constitute a bug. 
Both this issues ultimately are caused but not well defined behaviour of DataParallel with respect to stateful modules, of which batchnorm and recurrent nets with cudnn backend are examples. 

cc @ngimel @albanD @mruberry

**Comment by thatguymike** [comment]
This does seem bad.  For large networks like resnet50 where you have small batches, each GPU has a poor estimate of the norm, especially at the beginning of training.  We have found that allowing each GPU to run it's own batch norm can be more stable and better, calculating a total batch norm across all GPUs.  Final convergence is likely not impacted much since the final norm will work itself out after a few 10s of epochs.  

But I agree with @ngimel that this looks like a bug in general for anything with state even if we may argue about batch norm specifically.

**Comment by ngimel** [comment]
It would show up in testing only, no impact on training whatsoever (running vars are not used in training).

**Comment by colesbury** [comment]
1. It's probably simpler to disallow specifying the same device multiple times in DataParallel. I'm not sure there are sufficient use cases to warrant supporting it.

2. It would be better to accumulate running_mean/var, but this requires autograd to support "backward" ops on things that aren't optimized by SGD.

**Comment by ngimel** [comment]
2. I agree that accumulating running_mean/var across GPUs is not trivial in the current autograd model. But even if we decide not to accumulate them (arguably, over many epochs even the master GPU will see enough different samples to come up with good running estimates), current implementation broadcasting them to all the GPUs only to immediately discard whatever updates other GPUs made to them makes no sense. May be they should be not buffers, but some members of module __dict__, like, e.g. random states in rnns. 
Also, I think module replication in DataParallel should be better documented. Random states in rnn work now only because __dict__ is shallow copied in replicate, it would break if __dict__ were deep copied. But what if some module expects __dict__ to be deep copied to work properly? There's nothing in the docs to say that it wouldn't work.

**Comment by apaszke** [comment]
Shallow copy is mentioned in the last paragraph of the [`DataParallel` docs](http://pytorch.org/docs/nn.html#torch.nn.DataParallel).

**Comment by ngimel** [comment]
I read it as referring to inputs, not to module attributes.

**Comment by colesbury** [comment]
2. We could replace them with "dummy" buffers on the correct GPU during training. I'm not sure it will save enough time to make it a worthwhile optimization. At one point, we didn't broadcast the running_mean/var which led to other replicas using the original buffers (on the wrong GPU) which slowed everything down.

**Comment by ngimel** [comment]
Yes, that's what we had for random states in rnn, that's why rnn module now has a dict with rng state per GPU. Broadcasting buffers is ~1.5% of the runtime for resnet with 64/GPU, would be more for smaller batch sizes or future architectures.

**Comment by zhanghang1989** [comment]
Is there any update for this issue? It seems to be a serious problem for segmentation task, since the batch size is relatively small.

**Comment by zhanghang1989** [comment]
I fixed it by myself. Add this function to dataparallel table and call it at the end of forward. I hope it is helpful.

```python
    def _aggregate_bn(self, replicas):
        lock = threading.Lock()
        mean, var = {}, {}
        for k in range(len(self.device_ids)):
            mean[k] = {}
            var[k] = {}
        # subworker for individual gpu calculating the mean and var
        def _worker(k, model):
            i = 0
            #print('model:', model)
            for m in model.modules():
                classname = m.__class__.__name__ 
                if classname.find('BatchNorm2d') != -1:
                    try:
                        with lock:
                            mean[k][i] = m.running_mean
                            var[k][i] = m.running_var
                    except Exeption as e:
                        with lock:
                            mean[k][i] = e
                            var[k][i] = e
                    i += 1
        # across all the gpus
        threads = [threading.Thread(target=_worker,
                                    args=(i, model),
                                    )
                   for i, (model) in
                   enumerate(replicas)]
        for thread in threads:
            thread.start()
        for thread in threads:
            thread.join()
        # average all
        master_mean, master_var = {}, {}
        for i in range(len(mean[0])):
            kmean, kvar = 0, 0
            for k in range(len(self.device_ids)):
                kmean += mean[k][i].cpu()
                kvar += var[k][i].cpu()
            master_mean[i] = kmean / len(self.device_ids)
            master_var[i] = kvar / len(self.device_ids)

        # update the master model
        i = 0
        for m in self.module.modules():
            classname = m.__class__.__name__ 
            if classname.find('BatchNorm2d') != -1:
                m.running_mean = master_mean[i].cuda(0)
                m.running_var = master_var[i].cuda(0)
                i += 1
    
```

**Comment by zhanghang1989** [comment]
Is there a way to set the value of the register_buffer in `Dataparallel` ? I am able to gather the running_mean and running_variance for the batch norm layers, but I cannot set the values correclty. Any thoughts? I really appreciate the help!

**Comment by tingxueronghua** [comment]
Sorry, but I don't fully understand what you mean. When I use "register_buffer" to get a buffer and I update it during forward, do you mean that I cannot get the correct corresponding value on each GPU if I use DataParallel? Does it mean all the GPUs have to use the buffers on GPU:0?
